spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatch.java

```diff
@@ -115,18 +119,37 @@ private boolean parquetBatchReadsEnabled() {
   }
 
   private boolean orcOnly() {
-    return tasks().stream()
+    return taskGroups.stream()
         .allMatch(task -> !task.isDataTask() && onlyFileFormat(task, FileFormat.ORC));
   }
 
   private boolean orcBatchReadsEnabled() {
     return readConf.orcVectorizationEnabled()
         && // vectorization enabled
-        tasks().stream().noneMatch(TableScanUtil::hasDeletes); // no delete files
+        taskGroups.stream().noneMatch(TableScanUtil::hasDeletes); // no delete files
   }
 
   private boolean onlyFileFormat(CombinedScanTask task, FileFormat fileFormat) {
     return task.files().stream()
         .allMatch(fileScanTask -> fileScanTask.file().format().equals(fileFormat));
   }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
@bryanck, could you, please, re-run the benchmark in which you spotted the performance regression? I think the current logic should make `BatchScanExec` behave correctly.

Reply:
I ran the benchmark and this change doesn't have any negative impact on performance, so LGTM!

Reply:
One note, I backported this to Spark 3.2 and ran it with that, as Spark 3.3 has a performance regression (https://issues.apache.org/jira/browse/SPARK-40703)

Reply:
Thank you, @bryanck! It is nice to see Chao already looking into that. Let's see what was the root cause.